webapp/views.py

```diff
@@ -767,6 +767,116 @@ def post_advantage_subscriptions(preview):
     return flask.jsonify(purchase), 200
 
 
+def cancel_advantage_subscriptions():
+    api_url = flask.current_app.config["CONTRACTS_LIVE_API_URL"]
+
+    if flask.request.args.get("test_backend", False):
+        api_url = flask.current_app.config["CONTRACTS_TEST_API_URL"]
+
+    user_token = flask.session.get("authentication_token")
+
+    if user_info(flask.session):
+        advantage = AdvantageContracts(
+            session,
+            user_token,
+            token_type=("Macaroon" if user_token else "Bearer"),
+            api_url=api_url,
+        )
+    else:
+        return flask.jsonify({"error": "authentication required"}), 401
+
+    payload = flask.request.json
+
+    if not (
+        payload.get("account_id")
+        and payload.get("previous_purchase_id")
+        and payload.get("products")
+    ):
+        return flask.jsonify({"error": "bad request"}), 400
+
+    account_id = payload.get("account_id")
+    previous_purchase_id = payload.get("previous_purchase_id")
+
+    try:
+        monthly_subscriptions = (
+            advantage.get_account_subscriptions_for_marketplace(
+                account_id=account_id,
+                marketplace="canonical-ua",
+                filters={"status": "active", "period": "monthly"},
+            )
+        )
+    except HTTPError:
+        flask.current_app.extensions["sentry"].captureException(
+            extra={"payload": payload}
+        )
+        return (
+            flask.jsonify(
+                {"error": "could not retrieve account subscriptions"}
+            ),
+            500,
+        )
+
+    if not monthly_subscriptions:
+        return flask.jsonify({"error": "no monthly subscriptions found"}), 400
+
+    monthly_subscription = monthly_subscriptions.get("subscriptions")[0]
+
+    purchase_request = {
+        "accountID": account_id,
+        "purchaseItems": [
+            {
+                "productListingID": product["product_listing_id"],
+                "metric": "active-machines",
+                "value": 0,
+                "delete": True,
+            }
+            for product in payload.get("products")
+        ],
+        "previousPurchaseID": previous_purchase_id,
+    }
+
+    try:
+        purchase = advantage.purchase_from_marketplace(
+            marketplace="canonical-ua", purchase_request=purchase_request
+        )
+    except HTTPError as http_error:
+        api_response = http_error.response.json()
+
+        if "cannot remove all subscription items" in api_response["message"]:
```

Comment:
I think this check could be done by `advantage.purchase_from_marketplace`, that could return a specific exception for this case, like `CannotCancelLastContractError`, that you check here and react accordingly. What do you think?

Reply:
Good idea. Made the change.